libraries/chain/controller.cpp

```diff
@@ -2983,6 +2997,30 @@ bool controller::all_subjective_mitigations_disabled()const {
    return my->conf.disable_all_subjective_mitigations;
 }
 
+fc::optional<uint64_t> controller::convert_exception_to_error_code( const eosio_assert_code_exception& e ) {
+   const auto& logs = e.get_log();
+
+   if( logs.size() == 0 ) return {};
+
+   const auto msg = logs[0].get_message();
+
+   auto pos = msg.find( ": " );
+
+   if( pos == std::string::npos || (pos + 2) >= msg.size() ) return {};
+
+   pos += 2;
+
+   uint64_t error_code = 0;
+
+   try {
+      error_code = std::strtoull( msg.c_str() + pos, nullptr, 10 );
+   } catch( ... ) {
+      return {};
+   }
+
+   return error_code;
+}
+
```

Comment:
Instead of this, can we either manually derive `eosio_assert_code_exception` from `action_validate_exception` or create a new macro that adds an `error_code` attribute to the exception so we can just grab it directly instead of this string parsing.